apps/api/src/app/integrations/usecases/provider-usage-limits/provider-usage-limits.usecase.ts

```diff
@@ -0,0 +1,56 @@
+import { Inject, Injectable } from '@nestjs/common';
+import { IntegrationEntity, MessageRepository } from '@novu/dal';
+
+import { startOfMonth, endOfMonth } from 'date-fns';
+import { ANALYTICS_SERVICE } from '../../../shared/shared.module';
+import { AnalyticsService } from '@novu/application-generic';
+
+import { ProviderUsageLimitsCommand } from './provider-usage-limits.command';
+
+@Injectable()
+export class ProviderUsageLimits {
+  constructor(
+    private messageRepository: MessageRepository,
+    @Inject(ANALYTICS_SERVICE) private analyticsService: AnalyticsService
+  ) {}
+
+  static MAX_NOVU_INTEGRATION_MAIL_REQUESTS = parseInt(process.env.MAX_NOVU_INTEGRATION_MAIL_REQUESTS || '300', 10);
+
+  async execute(command: ProviderUsageLimitsCommand): Promise<number> {
+    return this.getMessageCount(command.organizationId, command.environmentId, command.providerId);
+  }
+
+  private async getMessageCount(organizationId: string, environmentId: string, providerId: string): Promise<number> {
+    return await this.messageRepository.count({
+      _organizationId: organizationId,
+      providerId: providerId,
+      createdAt: { $gte: startOfMonth(new Date()), $lte: endOfMonth(new Date()) },
+    });
+  }
+
+  async ensureLimitNotReached(integration: IntegrationEntity) {
+    if (!integration.limits) return;
+
+    const { softLimit, hardLimit } = integration.limits;
```

Comment:
What is soft limit for can you explain the usecase for it?

Reply:
The idea here is, if we have one limit then after reaching the limit we stop sending messages, instead of that if we have soft limit, after crossing soft limit we can just give warning but still send messages, after reaching hard limit we stop sending message, like soft limit is a warning and they can extend limit if needed after receiving this warning.